### ISSUE TITLE
Parallel chains hanging in IPython Notebook

### ISSUE THREAD

**Issue by kforeman** [stan-dev/pystan#142]
With 2.5.0.2 when I run multiple chains in parallel inside of an IPython Notebook, it hangs after sampling has finished. It gets stuck on `sampling()`, but checking the terminal for stdout shows that all chains have completed. If I interrupt the kernel I get the following error:

```
Exception in thread Thread-10:
Traceback (most recent call last):
  File "/Users/kforeman/anaconda/python.app/Contents/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/Users/kforeman/anaconda/lib/python2.7/site-packages/zmq/utils/garbage.py", line 46, in run
    s.bind(self.gc.url)
  File "socket.pyx", line 444, in zmq.backend.cython.socket.Socket.bind (zmq/backend/cython/socket.c:4076)
  File "checkrc.pxd", line 21, in zmq.backend.cython.checkrc._check_rc (zmq/backend/cython/socket.c:6032)
ZMQError: Address already in use
```

Seems like it may be related to [this issue](https://github.com/ipython/ipython/issues/6109) which implies that using Python's `multiprocessing` module from within IPython can cause issues. So maybe this is just an inherent issue with using multiprocessing from within IPython, but the strange thing is I had previously run multiple chains in an IPython notebook in 2.4.0.3.

**Comment by ariddell** [comment]
Thanks for this report. Does it work if you disable parallel sampling (set `n_jobs` to 1)?

It would be great to have pystan work in IPython notebooks. If multiprocessing is truly broken then maybe pystan should detect if the user is using an IPython notebook? (It does look like there's a
workaround for Python 3.4, which is nice).

**Comment by ariddell** [comment]
It looks like there are a lot of eyes on this problem. Hopefully a backport of the solution from 3.4 to Python 2.7 will arrive.

**Comment by kforeman** [comment]
Yes @ariddell `n_jobs=1` has been working fine for sampling multiple chains in serial. Couldn't find any indications in the python.org bug tracker that this is on the docket for a backport, so maybe this will be the thing that finally gets me to jump to 3.x. 

There's also hints that [billiard](https://github.com/celery/billiard/search?q=set_start_method) has backported this, but that would require using their `multiprocessing` fork. Supporting an optional dependency when `multiprocessing` is built in and works on 3.4 doesn't seem worth it.

**Comment by ariddell** [comment]
If you can confirm that everything works under IPython notebook 3.4 perhaps we should add a warning that detects IPython notebook usage under 2.7 and tells the user to consider upgrading or use `n_jobs=1`? Usability is vital, IMHO.

**Comment by kforeman** [comment]
@ariddell quite busy the next two weeks, but have put a note on my calendar to try setting up 3.4 and testing this the week of Feb 17.

**Comment by ariddell** [comment]
I'm going to close this for lack of activity. Please reopen if it's still a problem.

**Comment by mjwillson** [comment]
I'm seeing this sometimes, but not always, when running in an iPython notebook with Python 3.4. I haven't been able to pin down exactly how to replicate, but it's a similar "ZMQError: Address already in use" error. Will update if I track it down.

Incidentally is multiprocessing definitely necessary here? If the real work is happening in C++ code that doesn't touch Python objects, and there's no shared state between the chains, you should be able to release the GIL (means sprinkling a lot of nogil annotations into your cython .pxd's, and a "with nogil:" block around the entry point into it, and then use concurrent.futures.ThreadPoolExecutor to run it in different threads in the same process.

**Comment by ariddell** [comment]
Hi @mjwillson. I think you're right. I'm not sure how much ``nogil`` sprinkling is needed but it's worth checking out in any event.

**Comment by mjwillson** [comment]
IIRC Cython makes you sprinkle quite a bit, since within a nogil function you can only call other nogil functions. Still it's a fairly routine process.

**Comment by ariddell** [comment]
That's what I recall as well.  You're welcome to contribute a PR if you'd like. Otherwise I may get to this sometime this week.

**Comment by ariddell** [comment]
@mjwillson do you have an example where Cython (w/ nogil) is used with ThreadPoolExecutor? I can't seem to get my toy example to work (faster).

**Comment by mjwillson** [comment]
I'm afraid I don't have any public code examples, but I did get a significant speedup using this combination on a project here.

I was using Python 3.4, ThreadPoolExecutor.map with one worker per CPU core.

Worth checking that a significant proportion of time is actually spent in nogil code (could check with a profiler in a single-threaded setup first). And that the calls into the nogil code are happening inside a cython "with nogil:" block...

**Comment by mjwillson** [comment]
As another option there's also cython's OpenMP-based parallel, which I forgot about: http://docs.cython.org/src/userguide/parallelism.html -- haven't used this recently myself though

**Comment by ariddell** [comment]
I figured it out. There was some Python object in there.

It seems like wrapping stuff initially in a ``with nogil:`` block is useful for identifying lingering Python objects (even if the ``with nogil:`` block itself is unnecessary -- i.e., all your functions are declared to be ``nogil``)

**Comment by mjwillson** [comment]
Yeah, it's quite handy!

The 'with nogil' block is still necessary though, even if all the functions are declared as nogil. From http://docs.cython.org/src/userguide/external_C_code.html#declaring-a-function-as-callable-without-the-gil 
" The nogil function annotation declares that it is safe to call the function without the GIL. It is perfectly allowed to execute it while holding the GIL. The function does not in itself release the GIL if it is held by the caller."

**Comment by ariddell** [comment]
Forget what I said earlier about figuring it out.

Thanks for this clarification. I'll keep working at it. It would be very helpful to see one example where this works with pool.map (either from ``ThreadPoolExecutor.map`` or from ``multiprocessing.pool import ThreadPool``). It seems like you can't release the gil in this case since you're assigning to a Python object.

**Comment by ariddell** [comment]
OK. I got it using more than one core with ``joblib``. This is going to take some testing!

**Comment by mjwillson** [comment]
OK here's a quick sketch of what my code for this looks like -- YMMV depending on what datatypes if any need to go in and out of the CPP code, but hopefully gives you the idea:

pxd header:

```cython
cdef extern from "do_some_work.hpp" namespace "do_some_work":
    cdef cppclass CppArgDatatype:
        # ...
    cdef cppclass CppResultDatatype:
        # ...

    cdef CppResultDatatype native_do_some_work(CppArgDatatype arg) nogil
```


pyx cython
```cython
def do_some_work(input_python_object):

    cdef CppArgDatatype args_for_native_code = # do some stuff here to populate this from input_python_object

    cdef CppResultDatatype result_from_native_code

    with nogil:
        result_from_native_code = native_do_some_work(args_for_native_code)

    result = # do some stuff here if necessary to create a python object from result_from_native_code
    return result
```

python
```python
tpe = ThreadPoolExecutor(multiprocessing.cpu_count())
tpe.map(do_some_work, input_python_objects)
```

**Comment by ariddell** [comment]
Thanks, I got it working.

**Comment by ariddell** [comment]
I'm getting segfaults right now in my early attempts to get this working. There's no question in my mind that I should be using threads (with the gil released) instead of multiprocessing. Thanks to @mjwillson for making this clear to me. I've created issue #179 to track this.

**Comment by mjwillson** [comment]
Ah bum. I had kinda assumed that the C++ code itself is threadsafe -- if it's not then you might be stuck with multiprocessing :( I wonder if the other R bindings implement this via forking or threads?

**Comment by mjwillson** [comment]
https://github.com/stan-dev/stan/wiki/Model-Concept says "Stan models themselves are thread safe after construction. Operations on models involving auto-diff may not be thread safe." -- so you might need to make a (deep?) copy of the model for each thread

**Comment by ariddell** [comment]
That's worth a shot. I'll look into it.

**Comment by ariddell** [comment]
Closing as a duplicate of #179
